lib/galaxy/objectstore/cloud.py

```diff
@@ -0,0 +1,611 @@
+                # the limit - maybe delete additional #%?
+                # For now, delete enough to leave at least 10% of the total cache free
+                delete_this_much = total_size - cache_limit
+                self.__clean_cache(file_list, delete_this_much)
+            self.sleeper.sleep(30)  # Test cache size every 30 seconds?
+
+    def __clean_cache(self, file_list, delete_this_much):
+        """ Keep deleting files from the file_list until the size of the deleted
+        files is greater than the value in delete_this_much parameter.
+
+        :type file_list: list
+        :param file_list: List of candidate files that can be deleted. This method
+            will start deleting files from the beginning of the list so the list
+            should be sorted accordingly. The list must contains 3-element tuples,
+            positioned as follows: position 0 holds file last accessed timestamp
+            (as time.struct_time), position 1 holds file path, and position 2 has
+            file size (e.g., (<access time>, /mnt/data/dataset_1.dat), 472394)
+
+        :type delete_this_much: int
+        :param delete_this_much: Total size of files, in bytes, that should be deleted.
+        """
+        # Keep deleting datasets from file_list until deleted_amount does not
+        # exceed delete_this_much; start deleting from the front of the file list,
+        # which assumes the oldest files come first on the list.
+        deleted_amount = 0
+        for entry in enumerate(file_list):
+            if deleted_amount < delete_this_much:
+                deleted_amount += entry[2]
+                os.remove(entry[1])
+                # Debugging code for printing deleted files' stats
+                # folder, file_name = os.path.split(f[1])
+                # file_date = time.strftime("%m/%d/%y %H:%M:%S", f[0])
+                # log.debug("%s. %-25s %s, size %s (deleted %s/%s)" \
+                #     % (i, file_name, convert_bytes(f[2]), file_date, \
+                #     convert_bytes(deleted_amount), convert_bytes(delete_this_much)))
+            else:
+                log.debug("Cache cleaning done. Total space freed: %s", convert_bytes(deleted_amount))
+                return
+
+    def _get_bucket(self, bucket_name):
+        """ Sometimes a handle to a bucket is not established right away so try
+        it a few times. Raise error if connection is not established. """
+        for i in range(5):
+            try:
+                bucket = self.conn.object_store.get(bucket_name)
+                if bucket is None:
+                    log.debug("Bucket not found, creating a bucket with handle '%s'", bucket_name)
+                    bucket = self.conn.object_store.create(bucket_name)
+                log.debug("Using cloud object store with bucket '%s'", bucket.name)
+                return bucket
+            except S3ResponseError:
```

Comment:
I agree, and as explained in the previous comments, as a temporary patch, replaced `S3`/`boto`-specific exceptions with generic exceptions.